sdk/core/azure-core-v2/src/main/java/com/azure/v2/core/implementation/AccessTokenCache.java

```diff
@@ -31,137 +27,312 @@ public final class AccessTokenCache {
+            tokenRefresh = false;
+            fallback = cachedToken;
+        } else if (cachedToken == null || cachedToken.isExpired()) {
+            // no token to use
+            // refresh immediately
+            tokenRefresh = true;
+
+            // cache doesn't exist or expired, no fallback
+            fallback = null;
+        } else {
+            // token available, but close to expiry
+            if (now.isAfter(cache.getNextTokenRefresh())) {
+                // refresh immediately
+                tokenRefresh = true;
             } else {
-                // token available, but close to expiry
-                if (now.isAfter(cache.getNextTokenRefresh())) {
-                    // refresh immediately
-                    tokenRefresh = tokenSupplierSync;
-                } else {
-                    // still in timeout, do not refresh
-                    tokenRefresh = null;
-                }
-                // cache hasn't expired, ignore refresh error this time
-                fallback = cachedToken;
+                // still in timeout, do not refresh
+                tokenRefresh = false;
             }
+            // cache hasn't expired, ignore refresh error this time
+            fallback = cachedToken;
+        }
 
-            try {
-                if (tokenRefresh != null) {
-                    AccessToken token = tokenRefresh.get();
-                    buildTokenRefreshLog(LogLevel.INFORMATIONAL, cachedToken, now).log("Acquired a new access token.");
-                    OffsetDateTime nextTokenRefreshTime = OffsetDateTime.now().plus(REFRESH_DELAY);
-                    AccessTokenCacheInfo updatedInfo = new AccessTokenCacheInfo(token, nextTokenRefreshTime);
-                    this.cacheInfo.set(updatedInfo);
-                    return token;
-                } else {
-                    return fallback;
-                }
-            } catch (Throwable error) {
-                buildTokenRefreshLog(LogLevel.ERROR, cachedToken, now).log("Failed to acquire a new access token.",
-                    error);
-                OffsetDateTime nextTokenRefreshTime = OffsetDateTime.now();
-                AccessTokenCacheInfo updatedInfo = new AccessTokenCacheInfo(cachedToken, nextTokenRefreshTime);
-                this.cacheInfo.set(updatedInfo);
-                if (fallback != null) {
-                    return fallback;
-                }
-                throw error;
+        try {
+            if (tokenRefresh) {
+                AccessToken token = getToken();
+                logTokenRefresh(LogLevel.VERBOSE, cachedToken, now, "Acquired a new access token.");
+                OffsetDateTime nextTokenRefreshTime = OffsetDateTime.now().plus(REFRESH_DELAY);
+                AccessTokenCacheInfo updatedInfo = new AccessTokenCacheInfo(token, nextTokenRefreshTime);
+                this.cacheInfo = updatedInfo;
+                return token;
+            } else {
+                return fallback;
```

Comment:
Can this be broken down and moved to utils for the multiple if else's refreshIfStaleAndFetch or such?

Reply:
refactored this within class, util is not required.